### PR TITLE
[Kestrel] Allow HTTP/2 stream drain timeout during trailers

### PR DESCRIPTION
**Pull request by halter73**
aspnet/KestrelHttpServer#3077

### CODE REVIEW

**Comment by Tratcher** [comment]
Understated much? You redesigned the whole connection state management system.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 1]
```diff
@@ -1041,66 +966,103 @@ private void AbortStream(int streamId, IOException error)
...
+
+                    // If we're awaiting headers, either a new stream will be started, or there will be a connection
+                    // error possibly due to a request header timeout, so no need to start a keep-alive timeout.
+                    Debug.Assert(TimeoutControl.TimerReason == TimeoutReason.RequestHeaders ||
+                        TimeoutControl.TimerReason == TimeoutReason.KeepAlive);
                 }
             }
         }
 
         void IRequestProcessor.Tick(DateTimeOffset now)
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: Remove Tick

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 2]
```diff
@@ -1041,66 +966,103 @@ private void AbortStream(int streamId, IOException error)
             }
         }
 
-        void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
+        void IHttp2StreamLifetimeHandler.OnStreamCompleted(Http2Stream stream)
+        {
+            _completedStreams.Enqueue(stream);
+            Input.CancelPendingRead();
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: This is fishy and at least needs a comment.

**Review by Tratcher — COMMENTED** [review]
What were the benchmark results?

**Comment by natemcmaster** [comment]
Looks like this PR has gone stale. Should this be merged or closed?

**Comment by halter73** [comment]
I need to address some feedback and merge this. I'll do this tomorrow.

**Comment by halter73** [comment]
⬆️📅 I wound up going with int.MaxValue for the last stream id in abortive goaway frames.

**Review by Tratcher — COMMENTED** [review]
Any word on the benchmarks?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**